lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,712 @@
+-- @param number row Row number for the top left corner of the widget
+-- @param number col Column number for the top left corner of the widget
+-- @tparam number row_span The number of rows the widget spans (default to 1)
+-- @tparam number col_span The number of columns the widget spans (default to 1)
+-- @treturn table of index of widget_table
+local function find_widgets_by_coordinates(widgets_table, row, col, orig_row_span, orig_col_span)
+    local row_span = (orig_row_span and orig_row_span > 0) and orig_row_span or 1
+    local col_span = (orig_col_span and orig_col_span > 0) and orig_col_span or 1
+    local ret = {}
+    for index, data in ipairs(widgets_table) do
+        -- If one rectangular widget is on left side of other
+        local test_horizontal = not (row > data.row + data.row_span - 1 or data.row > row + row_span - 1)
+        -- If one rectangular widget is above other
+        local test_vertical   = not (col > data.col + data.col_span - 1 or data.col > col + col_span - 1)
+        if test_horizontal and test_vertical then
+            table.insert(ret, index)
+        end
+    end
+    table.sort(ret)
+    return #ret > 0 and ret or nil
+end
+
+--- Get (and update) the number of rows and columns occupied by the widgets in the grid
+-- @treturn number,number the number of rows and columns
+function grid:get_shape()
+    local num_rows, num_cols = 0, 0
+    if self._private.forced_num_rows then
+        num_rows = self._private.forced_num_rows
+    end
+    if self._private.forced_num_cols then
+        num_cols = self._private.forced_num_cols
+    end
+
+    for _, data in ipairs(self._private.widgets) do
+        num_rows = math.max(num_rows, data.row + data.row_span - 1)
+        num_cols = math.max(num_cols, data.col + data.col_span - 1)
+    end
+    self._private.num_rows = num_rows
+    self._private.num_cols = num_cols
+    return num_rows, num_cols
+end
+
+
+--- Add a widget to the grid layout at specific coordinate
+--@DOC_wibox_layout_grid_add_EXAMPLE@
+-- @param child Widget that should be added
+-- @param number row Row number for the top left corner of the widget
+-- @param number col Column number for the top left corner of the widget
+-- @tparam number row_span The number of rows the widget spans (default to 1)
+-- @tparam number col_span The number of columns the widget spans (default to 1)
+-- @treturn boolean index If the operation is successful
+function grid:add_widget_at(child, row, col, row_span, col_span)
+    local rspan = (row_span and row_span > 0) and row_span or 1
+    local cspan = (col_span and col_span > 0) and col_span or 1
+
+    -- check if the object is a widget
+    base.check_widget(child)
+
+    -- test if the new widget superpose with existing ones
+    local superpose = find_widgets_by_coordinates(self._private.widgets, row, col, rspan, cspan)
+    if superpose then
+        -- TODO: add message, remove other widgets?
```

Comment:
This TODO should be handled somehow. Personally, I'm fine with rejecting the addition.